### PR TITLE
browser: v1.8, a secure chat button for every profile

### PR DESCRIPTION
**Pull request by shazow**
## What's new:
- Rebranding from "Keybase for Reddit" to just "Keybase"
- Keybase Chat buttons on ALL the profiles!
  - [x] Hacker News
  - [x] Twitter
  - [x] Github
  - [x] Keybase
  - [x] Facebook
  - [x] Reddit

## Outstanding work:
- [x] Clean up injectors, maybe put them in a separate file and use a registry

### CODE REVIEW

**Comment by shazow** [comment]
@keybase/react-hackers Ready for review.

File: browser/js/content.js [hunk 1]
```diff
@@ -30,75 +27,40 @@ function init() {
...
+      const injectFn = profileInject[user.origin];
+      if (injectFn !== undefined) {
+        injectFn(user);
+      }
+    }
+  });
 }
 window.addEventListener('load', init);
 
-const checkThread = /^\/r\/\w+\/comments\/\w+\//;
```
Comment (shazow) [inline browser/js/content.js]: Below chunk got moved into inject.js

File: browser/js/content.js [hunk 2]
```diff
@@ -407,19 +407,3 @@ function findParentByClass(el, className) {
   }
   return null;
 }
-
-// Convert a user input into a string that is safe for inlining into HTML.
-function safeHTML(s) {
```
Comment (shazow) [inline browser/js/content.js]: Moving into identities.js which is a prereq for content anyways.

File: browser/js/content.js [hunk 3]
```diff
@@ -5,10 +5,7 @@
 const asset = chrome.runtime.getURL;
 
 function init() {
-  // Passive queries?
-  chrome.storage.sync.get("profile-passive-queries", function(options) {
-    if (options["profile-passive-queries"] !== true) return; // Default false
-
+  chrome.storage.sync.get(function(options) {
```
Comment (shazow) [inline browser/js/content.js]: Grab all options at once rather than one at a time

File: browser/js/content.js [hunk 4]
```diff
@@ -30,75 +27,37 @@ function init() {
...
+      });
+    }
 
-  // Inject Reddit thread DOM changes?
-  if (location.hostname.endsWith('.reddit.com') && checkThread.test(location.pathname)) {
-    chrome.storage.sync.get("reddit-thread-reply", function(options) {
-      // Is thread replies enabled?
-      if (options["reddit-thread-reply"] === false) return; // Default true
+    // Inject Reddit replies?
+    if (options["reddit-thread-reply"] !== false) { // undefined defaults to true
```
Comment (chrisnojima) [inline browser/js/content.js]: do you really want !== false here? so null /undefined will go into this if check?

File: browser/js/inject.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+function installExistingButtons(user) {
+  const preinstalled = document.getElementsByClassName("keybase-chat-open");
+  if (!preinstalled.length) return false;
+
+  installChatButton(preinstalled, user);
+  return true;
+}
+
+function injectProfileChat(user) {
+  const injectFn = profileInject[user.origin];
```
Comment (chrisnojima) [inline browser/js/inject.js]: check for null user?

File: browser/js/inject.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    const container = document.querySelector(".vcard-names");
+    if (!container) return;
+
+    const button = renderProfileChatButton(user);
+    container.appendChild(button);
+  },
+
+  "hackernews": function hackernewsInjectProfile(user) {
+    const tables = document.getElementsByTagName("tbody");
+    const profileTable = tables[tables.length-1];
```
Comment (chrisnojima) [inline browser/js/inject.js]: check for tables null

File: browser/js/inject.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    // Add a "chat" button next to username
+    const userLink = profileTable.children[0].children[1];
```
Comment (chrisnojima) [inline browser/js/inject.js]: check profileTable null and children null

File: browser/js/inject.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+function renderProfileChatButton(user) {
+  const icon = document.createElement("img");
+  icon.src = `${asset("images/icon-keybase-logo-16.png")}`;
+  icon.srcset = `${asset("images/icon-keybase-logo-16@2x.png")} 2x, ${asset("images/icon-keybase-logo-16@3x.png")} 3x`;
+
+  const button = document.createElement("a");
+  button.className = "keybase-chat";
+  button.href = `keybase://${user.query()}/`;
```
Comment (chrisnojima) [inline browser/js/inject.js]: do we need to do some escaping here?

File: browser/js/inject.js [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+  return button;
+}
+
+const redditCheckThread = /^\/r\/\w+\/comments\/\w+\//;
+function redditInjectThread(parent) {
+  // /r/<subreddit>/comments/<id>/<slug>
+  for (const c of parent.getElementsByClassName("comment")) {
+    const author = safeHTML(c.getAttribute("data-author"));
+    if (author == "") continue; // Empty
+    const buttons = c.getElementsByClassName("buttons")[0];
```
Comment (chrisnojima) [inline browser/js/inject.js]: null check

File: browser/js/content.js (hunk 4 above)
Comment (shazow) [inline browser/js/content.js]: Yes, the way the Chrome sync options work, the keys are not present by default so they're undefined). In this case we want reddit-thread-reply stuff to be enabled by default so undefined defaults to true (per the comment next to it).

**Comment by shazow** [comment]
Re: null checks in general, per chat, we're going to lean towards exploding with an exception if the DOM changes in an unexpected way rather than silently failing until we have some reason/provision for recovering from errors.

**Pull request merged**